File: tests/support_scripts/config_type_consumer.ts
import type { PcbSvgConfigInput } from "../../src/ts/altium_cruncher_config/index.js";
import { decodeContract } from "../../src/ts/altium_cruncher_config/index.js";

const builtinResult = decodeContract("mco_builtin_result", {} as unknown);
if (builtinResult.status === "ok" && builtinResult.op === "pcbdoc.add_via") {
  const position: number[] = builtinResult.outputs.position_mils;
  // @ts-expect-error A point cannot become an untyped nested dictionary.
  const invalidPosition: string = builtinResult.outputs.position_mils;
  void [position, invalidPosition];
}

declare const config: PcbSvgConfigInput;
const label: boolean | string | number | null | undefined = config.components?.J1?.show_designator;
const extension: unknown = config.global?.styles?.custom?.["vendor"];
const input: PcbSvgConfigInput = {
  components: { J1: { show_designator: false } },
  global: { styles: { custom: { vendor: { nested: 1 } } } },
};
// Record values must retain their type, not degrade to an empty interface/unknown.
// @ts-expect-error An array is not a component override.
const bad: PcbSvgConfigInput = { components: { J1: [] } };
void [label, extension, input, bad];

import type { ProjectSkeletonConfigInput, PcbdocCreateConfigInput } from "../../src/ts/altium_cruncher_config/index.js";
import type { FlatMechanicalPair } from "../../src/ts/altium_cruncher_config/generated/pcbdoc-create.js";
const project: ProjectSkeletonConfigInput = {
  schema: "altium_cruncher.project_skeleton.a0", project: {file: "a.PrjPcb", parameters: {Count: 2}},
  schematics: [], pcb: null, editor: {expanded: false},
};
const pcb: PcbdocCreateConfigInput = {
  schema: "altium_cruncher.pcbdoc.create.config.a0", file: "a.PcbDoc",
  mechanical_layer_kinds: [{layer: "MECHANICAL2", kind: 0}],
};
// @ts-expect-error A creation config requires a file.
const missingFile: PcbdocCreateConfigInput = {schema: "altium_cruncher.pcbdoc.create.config.a0"};
// @ts-expect-error Presence of top selects grouped form, never a flat pair.
const badPair: FlatMechanicalPair = {layer_1: "MECHANICAL10", layer_2: "MECHANICAL11", top: {}};
void [project, pcb, missingFile, badPair];

import type { BuiltinOperation, McoDocument } from "../../src/ts/altium_cruncher_config/index.js";
declare const operation: BuiltinOperation;
if (operation.op === "pcbdoc.add_text") {
  const height: number = operation.args.height_mils;
  const text: string = operation.args.text;
  void [height, text];
}
// @ts-expect-error add_text requires height_mils.
const missingHeight: BuiltinOperation = {op: "pcbdoc.add_text", args: {file: "a.PcbDoc", text: "Hi", position_mils: [0, 0]}};
// @ts-expect-error The container requires its ordered operation list.
const missingOperations: McoDocument = {schema: "altium_cruncher.mco.a0"};
void [missingHeight, missingOperations];

File: tests/support_scripts/config_type_consumer.py
"""Static consumer of generated requiredness, record values and open styles."""

from typing import assert_type
from altium_cruncher.contracts.generated.public import decode_contract

builtin_result = decode_contract("mco_builtin_result", {})
if builtin_result["status"] == "ok" and builtin_result["op"] == "pcbdoc.add_via":
    assert_type(builtin_result["outputs"]["position_mils"], list[float])

from altium_cruncher.contracts.generated.pcb_svg_config import (
    ComponentOverride,
    IllustrationStyle,
    PcbSvgConfigInput,
    StyleTable,
)

component = ComponentOverride(show_designator=False)
style = IllustrationStyle(vendor_option=True)
styles = StyleTable(custom={"vendor": {"nested": 1}}, illustration=style)
config: PcbSvgConfigInput = {
    "components": {"J1": component},
    "global": {"styles": styles},
}

from altium_cruncher.contracts.generated.project_skeleton_config import (
    ProjectSkeletonConfigInput,
)
from altium_cruncher.contracts.generated.pcbdoc_create_config import (
    PcbdocCreateConfigInput,
    FlatMechanicalPair,
)

project: ProjectSkeletonConfigInput = {
    "schema": "altium_cruncher.project_skeleton.a0",
    "project": {"file": "a.PrjPcb", "parameters": {"Count": 2}},
    "schematics": [],
    "pcb": None,
    "editor": {"expanded": False},
}
pair = FlatMechanicalPair(
    layer_1="MECHANICAL10", layer_2="MECHANICAL11", pair_index=None
)
pcb: PcbdocCreateConfigInput = {
    "schema": "altium_cruncher.pcbdoc.create.config.a0",
    "file": "a.PcbDoc",
    "mechanical_layer_pairs": [pair],
    "mechanical_layer_kinds": [{"layer": "MECHANICAL2", "kind": 0}],
}

from altium_cruncher.contracts.generated.mco_input import BuiltinOperation, McoDocument

operation: BuiltinOperation = {
    "op": "pcbdoc.add_text",
    "args": {
        "file": "a.PcbDoc",
        "text": "Hi",
        "position_mils": [0, 0],
        "height_mils": 40,
    },
}
document: McoDocument = {"operations": [operation]}
